File: fortran_mpi_pi_dir/pi.f90
program pi

  use ISO_FORTRAN_ENV
  use mpi

  implicit none

  double precision              :: step, x, s, finalsum, mypi, start, stop
  integer(kind=int64)           :: num_steps, i, mymax, mymin
  character(len=:), allocatable :: a
  integer                       :: argl, rank, ierr, numprocs

  call MPI_INIT(ierr)
  call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
  call MPI_COMM_SIZE(MPI_COMM_WORLD, numprocs, ierr)

  num_steps = 1000000000

! Get command line args (Fortran 2003 standard)
  if (command_argument_count() > 0) then
     call get_command_argument(1, length=argl)
     allocate(character(argl) :: a)
     call get_command_argument(1, a)
     read(a,*) num_steps
  end if

! Output start message

  if (rank == 0) then
    write(*,'(A)') "Calculating PI using:"
    write(*,'(A,1I16,A)') "                  ",num_steps, " slices"
    write(*,'(A,1I16,A)') "                  ",numprocs," process(s)"
  end if

! Initialise time counter and sum: set step size

  call MPI_BARRIER(MPI_COMM_WORLD, ierr)

  start = MPI_WTIME()
  s = 0d0
  step = 1.0d0 / num_steps

! Remember Fortran loops from 1
  mymin = ((rank * num_steps)/numprocs) + 1
  mymax = ((rank + 1) * num_steps)/numprocs

  do i = mymin, mymax
    x = (i - 0.5d0) * step
    s = s + 4.0d0 / (1.0d0 + x*x)
  end do

  call MPI_Reduce(s, finalsum, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD, ierr)

! Evaluate PI from the final sum value, and stop the clock

  mypi = finalsum * step

  call MPI_BARRIER(MPI_COMM_WORLD, ierr)
  stop =  MPI_WTIME()

! output value of PI and time taken
! note cpu_time is only specified as being microsecond res

  if (rank == 0) then
    write(*,'(A,1F12.10,A)') "Obtained value of PI: ", mypi
    write(*,'(A,1F12.5,A)') "Time taken:           ",(stop-start), " seconds"
  end if

  call MPI_FINALIZE(ierr)

end program pi

File: pascal_pi_dir/pi.pas
Program pi(output);

Uses sysutils;

{ Use am Int64 for numsteps because we want to use big numbers of steps. }
Var 
  step, x, sum, mypi, time, start, stop : Double;
  i, numsteps : Int64;

Begin
 
  numsteps := 1000000000;

{ Get numsteps from command line args. }
  if ParamCount > 0 then
    numsteps := StrToInt64(ParamStr(1));

  Writeln('Calculating PI using:');
  Writeln('  ', numsteps, ' slices');

{ Initialise time counter and sum. Set step size. }
  start := TimeStampToMSecs(DateTimeToTimeStamp(Now)); 

  sum := 0.0;
  step := 1.0/Double(numsteps);

  For i := 1 To numsteps Do
  
  Begin
    x := (Double(i) - 0.5) * step;
    sum := sum + 4.0/(1.0 + x*x);
  End;
  
{ Calculate pi from final value and stop clock. }
  mypi := sum * step;
  stop := TimeStampToMSecs(DateTimeToTimeStamp(Now)); 

{ Note: Highest resolution of this method is milliseconds.}
  time := (stop - start)/1000;

  Writeln('Obtained value of PI: ', FloatToStrf(mypi, fffixed, 2, 8));
  Writeln('Time taken: ', FloatToStrf(time, fffixed, 4, 4), ' seconds');

End.
